Scrape w3 for ipfs uris and write them to disk

njgheorghita opened the pull request:
## What was wrong?
CLI command needed to scrape blockchain for `VersionRelease` events - grab version release data - and write any emitted ipfs assets to disk

## How was it fixed?
Wrote `Scraper` class that takes a web3 instance and scrapes for `VersionRelease` data, writing any found data to disk.

#### Cute Animal Picture
![image](https://user-images.githubusercontent.com/9753150/56891148-cf1fcc00-6a7b-11e9-8287-15f2328b2c8a.png)

njgheorghita committed:
Write initial Scraper class

njgheorghita committed:
Update tests to work with multiple logs

njgheorghita commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,195 @@
+        logger: Logger = None,
+    ) -> None:
+        self.w3 = w3
+        # ws w3 instance needed since infura requires ws for certain jsonrpc calls
```
@kclowes - Infura only supports [certain jsonrpc](https://infura.io/docs) endpoints over ws (i.e. `eth_newFilter`) and others only over https (i.e. `eth_getBlock`).  I had a thought that it might be worth making web3 smart enough to direct the right calls to the relevant endpoints - and avoid having to use 2 different w3 instances - but it only applies to infura, so I'm not sure if it's worth it, just a thought.

njgheorghita replied:
@kclowes scratch that re: [this](https://github.com/ethpm/ethpm-cli/pull/3#discussion_r279393654) `local_filter_middleware` works great!

pipermerriam commented on ethpm_cli/constants.py:
```diff
@@ -1 +1,18 @@
+from eth_utils import keccak, to_hex
+from ethpm.constants import INFURA_API_KEY
+
+CONTENT_DIR = "ipfs_assets"
```
I think that this should make use of the [XDG spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) which would mean storing this data somewhere like `$HOME/.local/share/ethpm` or maybe `$HOME/.local/cache/ethpm` depending on the semantics of how we want to store this data.

You can see how we negotiate these directories [here in the trinity codebase](https://github.com/ethereum/trinity/blob/master/trinity/_utils/xdg.py)

*sidenote* I think that file could easily be extracted to a small python library to make it easier to generate those paths for different applications.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,195 @@
+        w3: Web3,
+        ws_w3: Web3 = None,
+        content_dir: Path = None,
+        logger: Logger = None,
```
I've come to see passing loggers around as an anti-pattern.  I'd recommend just setting it on the class.

```python
class Scraper:
    logger = logging.getLogger('ethpm.cli.scraper.Scraper')
```

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,195 @@
+        logger: Logger = None,
+    ) -> None:
+        self.w3 = w3
+        # websocket w3 instance needed since infura requires ws for certain jsonrpc calls
```
Take a look at this: https://github.com/ethereum/web3.py/blob/71b8bf03dc6d332dd97d8902a38ffab6f8b5a5ab/web3/middleware/filter.py#L284

This PR's functionality seems like the ideal use case for that middleware.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,195 @@
+        # websocket w3 instance needed since infura requires ws for certain jsonrpc calls
+        self.ws_w3 = ws_w3 or w3
+        self.version_release_data: OrderedDict[str, Any] = OrderedDict()
+        self.last_processed_block = 0
```
This statefulness gets bugs me a little, but I think that's mostly because I can see how we could do some small changes to remove it, but that may mean giving up on the scraper class, or at least extracting the functionality out of it.  Here's the gist.

```
class Config(NamedTuple):
    w3: Web3
    from_block: identifier  # maybe `None` signifies genesis
    to_block: identifier  # maybe `None` signifies to current


def process_uri(config, ipfs_uri) -> Set[Tuple[URI, Manifest]]:
    # download -> validate -> recurse-dependencies
    ...


def scrape_manifests_from_chain(config):
    # Low level function that is somewhat naive of statefulness of system.
    filter = w3.eth.filter(...)

    for log in filter:
        uri = get_uri_from_log(log)
        yield log, process_uri(config, uri)


def scrape(config, to_path: Path):
    # High level function that manages statefullness of system
    try:
        for log, manifests in scrape_chain(self.config):
            ...  
            # maybe write them to disk?
    except ...:
        # maybe save our progress (last processed log...)
```

pipermerriam reviewed (commented):
Incomplete review, will finish it out in a bit.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,195 @@
+            )
+
+        self.logger.info(
+            f"Block #:{self.last_processed_block} scraped. "
```
see https://github.com/ethereum/snake-charmers-tactical-manual/blob/master/style-guide.md#formatting-log-messages

pipermerriam reviewed (commented):
Going to wait to review the rest until later but wanted to ask for a more detailed write-up of what it scrapes and what data it writes to disk, and how that data is organized.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,195 @@
+from ethpm.utils.ipfs import extract_ipfs_path_from_uri, is_ipfs_uri
+from ethpm.utils.uri import is_supported_content_addressed_uri
+from web3 import Web3
+from web3._utils.events import get_event_data
```
Noting the import from a private API.  IIUC I think you should be able to define a contract factory that has the relevant event ABI and then use the contract APIs to pull logs that are properly parsed.

njgheorghita commented:
@pipermerriam Here's a little more on the structure of the data stored, it's more or less just the minimum I thought I needed to get things working
https://gist.github.com/njgheorghita/1ed9e6e061a5a84354cf08bc8d32812b

njgheorghita committed:
Update scraper to be stateless

pipermerriam reviewed (commented):
What would you say to dropping the complexity here a bit.  My main goal is to reduce the amount of data stored in JSON documents to be constant size (or very near constant).

Instead of storing all of the metadata, what if we only store the following:

- what blocks we've scraped either in the form of a singular range, or in the form of a list of ranges which we collapse when possible.
- a directory (or nested set of directories) where the contents can be stored and looked up by hash:
  - Something like `$XDG_CACHE/ethpm/ipfs/<first-two-hex-bytes-of-hash>/<second-two-hex-bytes-of-hash>/<remainder-of-hash>/manifest.json`
  -  Maybe store the metadata about the registry, package names, version, whatever alongside the `manifest.json`.

Basically trying to have the data be as self-describing as possible so that another tool that wants a file can look for it without having to understand the whole schema of the various JSON documents that hold metadata.  Also, trying to avoid having **any** large JSON documents.

njgheorghita committed:
Store scraped blocks as list of ranges in event_data.json

njgheorghita committed:
Update write to disk strategy

njgheorghita commented:
Updated ipfs assets dir structure 
```
- 'ipfs/'
    - 'chain_data.json'
    - 'Qm/'
        - 'be/'
            - ...
        - 'dv/'  (dir with first 2 unique bytes)
            - 'ZEW3AaUntDfFkcbdnYzeLAAeD4YFeixQsdmHF88T6Q/' (dir with remainder of hash)
                - 'QmdvZEW3AaUntDfFkcbdnYzeLAAeD4YFeixQsdmHF88T6Q' (file w/ name of entire hash, stores contents)
```
@pipermerriam  Thoughts on this structure? I eliminated storing metadata since it's not obviously necessary and tracking all the assets became fairly complex since we're not only storing manifests, but also their build dependencies and any contract files, documentation etc.. But, I can re-implement it if you think it's going to be important.

I've added a list of `"registry_addresses"` to the `"chain_data.json"` file. It violates your "keep json small" rule, but it shouldn't grow too quickly, and seems like that information could be useful. (This list of "active EthPM V2" registry addresses was the prior motivation for storing any metadata in the first place)

pipermerriam commented:
What about separating the scraping of registry addresses into a separate utility?  Feel free to punt on that and keep things as-is for now if it's useful to you.

For directory structure I'd go one level deeper with `Qm/dv/ZE/QMdvZEW3.....`  The extra level of namespacing is cheap and it is a huge PITA when you get a directory with way too many things in it because at a certain point `ls` stops working.

As for metadata, are we still scraping the files, just not storing information about what file is what?  As long as we're pulling the contract source files and dependency manifests I think we're :+1: .  Not pulling dependencies is probably a no-go since without the dependencies a package is likely useless.

njgheorghita commented:
> What about separating the scraping of registry addresses into a separate utility? Feel free to punt on that and keep things as-is for now if it's useful to you.

Sounds good - I'll open up a new pr to handle that task. 

> For directory structure I'd go one level deeper with Qm/dv/ZE/QMdvZEW3..... 

👍 

> As long as we're pulling the contract source files and dependency manifests I think we're 👍 . 

We're 👍

pipermerriam commented:
giving one last :eyes:

pipermerriam commented on __main__.py:
```diff
@@ -21,10 +28,40 @@ def get_logger():
+    w3 = setup_scraper()
+    ethpmcli_dir = get_xdg_ethpmcli_root()
+    scrape(w3, ethpmcli_dir, args.start_block)
+    logger.info("All blocks scraped up to # %d.", w3.eth.getBlock)
```
Maybe have scrape return the latest scraped block or even some sort of `ScrapeResult` data object to prevent a race condition where a new block comes in between `scrape` finishing and this reporting at which point the user might later use this log statement to determine which block to start from and miss any intermediate blocks.  Also, probably good to ensure the statement has both block number and hash (current statement looks like it dumps the whole block which seems like it'd be noisy)

pipermerriam commented on __main__.py:
```diff
@@ -53,25 +90,28 @@ def parse_arguments():
+            args.uri,
+            config.ethpm_dir,
         )
+    if args.command == "scrape":
```
Consider using this pattern (can be done in another PR).

https://github.com/ethereum/trinity/blob/9608a54c9985f60a32278b3115754f5fadf0cb6e/trinity/plugins/builtin/fix_unclean_shutdown/plugin.py#L28-L33

https://github.com/ethereum/trinity/blob/9608a54c9985f60a32278b3115754f5fadf0cb6e/trinity/bootstrap.py#L196-L197

Each cli command specifies it's own execution function so you can keep this area of the codebase clean of the individual business logic for each CLI command.

pipermerriam commented on __main__.py:
```diff
@@ -53,25 +90,28 @@ def parse_arguments():
-            "to see the list of available commands."
+            "%s is an invalid command. Use `ethpmcli --help` to see the list of available commands.",
+            args.command,
         )
```
This should probably be done via an [parser.error](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.error) API so that it exits with a non-zero exit code to ensure scripts don't interpret this as a happy path.

pipermerriam commented on ethpm_cli/_utils/various.py:
```diff
@@ -0,0 +1,13 @@
+
+
+@to_list
+def flatten(l: List[Any]) -> Iterable[Any]:
```
Curios if mypy allows this:

```
TValue = TypeVar("TValue")
TList = Sequence[Union[TValue, "TList"]]  # recursive definition...
```

This would in theory allow retention of the value type.  Not a big deal if it doesn't work.

njgheorghita replied:
Not supported *yet* ...
`Recursive types not fully supported yet, nested types replaced with "Any"`

pipermerriam commented on ethpm_cli/_utils/xdg.py:
```diff
@@ -0,0 +1,42 @@
+
+def get_home() -> Path:
+    try:
+        return Path(os.environ["HOME"])
```
Should probably use [`Path.home`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.home)

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+from eth_utils.toolz import assoc
+from ethpm.typing import URI, Address
+from ethpm.utils.backend import resolve_uri_contents
+from ethpm.utils.ipfs import extract_ipfs_path_from_uri, is_ipfs_uri
```
These need to eventually be exposed from the `ethpm` library as a public API.  Currently this is importing from private API namespace.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+
+    except BlockNotFoundError:
+        logger.info("No more blocks available.")
+        return None
```
I *think* convention is just a bare `return` since we aren't actually intending to return the `None` value, just that we're returning.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+    Scrapes VersionRelease event data starting from start_block.
+    """
+    chain_data_path = ethpm_dir / "chain_data.json"
+    try:
```
This `try` block looks like it contains more statements than necessary.  Can `BlockNotFoundError` come from any of these four statements?  If not, we should adjust to have only the statements which can raise be present in the `try` (even if you need to make more than one `try/except` block.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+    except BlockAlreadyScrapedError:
+        active_block = start_block + 1
+
+    scrape(w3, ethpm_dir, active_block)
```
This recursive approach is going to be problematic.  It will add a level of recursion for every already scraped block which will VERY quickly exceed the stack depth limit.  Should adjust this to use a loop and a counter and to break from the loop and enter the *real* scrape function once it finds the first unscraped block.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+        if start_block >= w3.eth.blockNumber:
+            raise BlockNotFoundError(
+                f"Block number: {start_block} not available on provided web3 "
+                f"instance with latest block number of {w3.eth.blockNumber}."
```
race condition here,  probably should pull off `w3.eth.blockNumber` to a local variable so that you don't encounter a situation where it changes between the two locations this property is accessed.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+            "scraped_blocks": [{"min": "0", "max": "0"}],
+        }
+        chain_data_path.write_text(f"{json.dumps(init_json, indent=4)}\n")
+        return 1
```
This conflates the two concepts of:

-  *ensure that the directory and file structure is present*
- *tell me which block I should start scraping from*

I recommend separating these into two different functions.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+@to_list
+def blocks_to_ranges(blocks_list: List[int]) -> Iterable[Dict[str, str]]:
+    """
+    Takes a list of block numbers, and returns them grouped into min/max ranges.
```
I *think* I understand what this says but I'm not sure.  Can you maybe provide a small code example in this docstring so it is illustrative?

Is this what it does?

https://github.com/ethereum/trinity/pull/569/files#diff-3c5a1b448f91c1258a0e43b661b3af54R16

njgheorghita replied:
yup! more or less, just formatted slightly differently

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+    ]
+    base_ipfs_uris = [uri for uri in all_manifest_uris if is_ipfs_uri(uri)]
+    nested_ipfs_uris = [pluck_ipfs_uris_from_manifest(uri) for uri in all_manifest_uris]
+    all_ipfs_uris = set(flatten(nested_ipfs_uris) + base_ipfs_uris)
```
This might be more appropriate done as `set(flatten(nested_ipfs_uris)).update(*base_ipfs_uris)`

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+def write_ipfs_uris_to_disk(
+    ethpm_dir: Path, manifests: Dict[Address, Dict[str, str]]
+) -> None:
+    ipfs_dir = ethpm_dir / "Qm"
```
I don't think this is ok.  While every IPFS URI you've worked with starts with this, it isn't guaranteed.  This prefix merely specifies the hash type and all implementations use SHA256.   Hard coding `Qm` is going to shoot us in the foot when an alternate hash-scheme shows up.

pipermerriam commented on ethpm_cli/scraper.py:
```diff
@@ -0,0 +1,215 @@
+            asset_dest_path = remainder_hash_dir / ipfs_hash
+            asset_dest_path.touch()
+            asset_dest_path.write_bytes(resolve_uri_contents(uri))
+            logger.info("%s written to disk.", uri)
```
Probably prudent to include the `path` in this message too..

pipermerriam commented on tests/core/ipfs/Qm/NQ/euwMDGJ7UiLaRjwzAoekcaKLp9TjiqeFdovj3syN1n/QmNQeuwMDGJ7UiLaRjwzAoekcaKLp9TjiqeFdovj3syN1n:
```diff
@@ -0,0 +1,23 @@
+pragma solidity ^0.4.24;
```
Looks like fixtures that should be zipped up to prevent accidentally modifying them in-place?

pipermerriam reviewed (approved):
Feel free to move this forward at whatever pace you need.  Few things in here that should be cleaned up but it's fine to punt them into tracking issues if you so desire.

njgheorghita committed:
Remove recursion from scraper
